### PR TITLE
If a fork condition is met we should go back to the confirm page (unless another fork condition is met)

### PR DESCRIPTION
**Pull request by easternbloc**
Say we have two conditions for a step:

```
'/letter-received': {
    fields: [
      'received',
      'delivery-date',
      'delivery-date-day',
      'delivery-date-month',
      'delivery-date-year',
      'no-letter'
    ],
    next: '/same-address',
    forks: [{
      target: '/letter-not-received',
      condition: {
        field: 'received',
        value: 'no'
      }
    }, {
      target: '/letter-lost',
      condition: {
        field: 'no-letter',
        value: 'true'
      }
    }]
  }
```

Currently if get to the edit (confirm page) then go back to that step to edit it (and hit a fork) it won't go down the fork it will just return to the edit page.

This is a bug as it should potentially end journeys for people or ask for extra information (on other steps).

Here is it broken:

And here is it with the base controller change in this PR:

### CODE REVIEW

**Comment by easternbloc** [comment]
@JoeChapman @daniel-ac-martin @timgent I'm not sure this is the best solution but there are cases where the fork lets you proceed back to the confirm page when it should instead push you to an exit page. The only worry is this will push you back down the form if you meet a fork that was ok... It's annoying because forks can be used for good and bad routes. :(

File: lib/base-controller.js [hunk 1]
```diff
@@ -26,24 +26,29 @@ function getErrorLength() {
...
     if (_.isFunction(fork.condition)) {
       if (fork.condition(req, res)) {
         next = req.baseUrl + fork.target;
+        forked = true;
       }
     }
     if (_.isPlainObject(fork.condition)) {
       if (fork.condition.value === req.form.values[fork.condition.field]) {
         next = req.baseUrl + fork.target;
+        forked = true;
```
Comment (JoeChapman) [inline lib/base-controller.js]: You could just return `next` here

File: lib/base-controller.js [hunk 2]
```diff
@@ -26,24 +26,29 @@ function getErrorLength() {
 BaseController.prototype.getNextStep = function getNextStep(req, res) {
   var next = Controller.prototype.getNextStep.apply(this, arguments);
   var forks = (this.options || {}).forks || [];
+  var forked = false;
 
   _.each(forks, function eachFork(fork) {
     if (_.isFunction(fork.condition)) {
       if (fork.condition(req, res)) {
         next = req.baseUrl + fork.target;
+        forked = true;
```
Comment (JoeChapman) [inline lib/base-controller.js]: You could just return `next` here

**Comment by JoeChapman** [comment]
@easternbloc please describe the problem you are trying to solve

File: lib/base-controller.js [hunk 3]
```diff
@@ -23,27 +23,42 @@ function getErrorLength() {
...
-    next = req.baseUrl === '/' ? this.confirmStep : req.baseUrl + this.confirmStep;
-  } else if (req.params.action === 'edit') {
-    next += '/edit';
+  Controller.prototype.saveValues.apply(this, arguments);
+};
+
+BaseController.prototype.getNextStep = function getNextStep(req) {
+  var next = Controller.prototype.getNextStep.apply(this, arguments);
+
+  if (req.sessionModel.get('fork')) {
```
Comment (JoeChapman) [inline lib/base-controller.js]: the tests are failing at ln 51 with "cannot call get of undefined"

File: lib/base-controller.js [hunk 4]
```diff
@@ -23,27 +23,42 @@ function getErrorLength() {
...
+BaseController.prototype.saveValues = function saveValues(req) {
   var forks = (this.options || {}).forks || [];
 
   _.each(forks, function eachFork(fork) {
-    if (_.isFunction(fork.condition)) {
-      if (fork.condition(req, res)) {
-        next = req.baseUrl + fork.target;
-      }
-    }
     if (_.isPlainObject(fork.condition)) {
```
Comment (JoeChapman) [inline lib/base-controller.js]: Is checking that the `fork.condition` is an object still necessary?

File: lib/base-controller.js [hunk 5]
```diff
@@ -23,27 +23,42 @@ function getErrorLength() {
   }
 }
 
-BaseController.prototype.getNextStep = function getNextStep(req, res) {
-  var next = Controller.prototype.getNextStep.apply(this, arguments);
+BaseController.prototype.saveValues = function saveValues(req) {
```
Comment (JoeChapman) [inline lib/base-controller.js]: Arrow function?

File: lib/base-controller.js [hunk 6]
```diff
@@ -23,27 +23,42 @@ function getErrorLength() {
...
   var forks = (this.options || {}).forks || [];
 
   _.each(forks, function eachFork(fork) {
```
Comment (JoeChapman) [inline lib/base-controller.js:31]: arrow function?

File: lib/base-controller.js [hunk 7]
```diff
@@ -23,27 +23,42 @@ function getErrorLength() {
...
-    if (_.isFunction(fork.condition)) {
-      if (fork.condition(req, res)) {
-        next = req.baseUrl + fork.target;
-      }
-    }
     if (_.isPlainObject(fork.condition)) {
       if (fork.condition.value === req.form.values[fork.condition.field]) {
```
Comment (JoeChapman) [inline lib/base-controller.js]: It might be a good idea to check that `fork.condition` is an object or accessing ``value` or `.field` will throw an error

File: lib/base-controller.js (hunk 6 above)
Comment (easternbloc) [inline lib/base-controller.js:31]: I had the entire class es6'd but then linting failed and i wanted to push the concept up. I think es6 should perhaps be another pr for this repo as it diverges from what i'm trying to do here.

File: lib/base-controller.js (hunk 5 above)
Comment (easternbloc) [inline lib/base-controller.js]: See below

File: lib/base-controller.js (hunk 3 above)
Comment (easternbloc) [inline lib/base-controller.js]: yup I need to fix the tests.

File: lib/base-controller.js (hunk 6 above)
Comment (JoeChapman) [inline lib/base-controller.js:31]: I think you're probably correct. We should make this bug fix available without requiring that users have es6 support

**Comment by JoeChapman** [comment]
@easternbloc hey dude, are you going to fix this?

**Comment by easternbloc** [comment]
@JoeChapman yes I just don't have time at the moment. It's here so i don't forget

**Comment by JoeChapman** [comment]
@easternbloc perhaps someone else should pick this up?

**Comment by easternbloc** [comment]
@JoeChapman I'll get round to it today I think.

**Comment by JoeChapman** [comment]
@easternbloc Documentation?

**Comment by easternbloc** [comment]
@JoeChapman will exist in the tests

**Comment by JoeChapman** [comment]
No, you need to update the README, it includes instructions for using the functional fork, and this removes the ability to use a function, no?

**Comment by easternbloc** [comment]
how about you let me actually finish the pr

**Comment by JoeChapman** [comment]
@easternbloc if only ;)

**Comment by easternbloc** [comment]
@JoeChapman

File: lib/base-controller.js [hunk 8]
```diff
@@ -26,27 +26,37 @@ function getErrorLength() {
...
+          next = req.baseUrl + fork.target;
+          forked = true;
+        }
       }
     }
   });
 
-  if (req.params.action === 'edit' && !this.options.continueOnEdit) {
+  if (forked) {
+    return next;
```
Comment (JoeChapman) [inline lib/base-controller.js:51]: Is this condition and statement necessary, can't you just return `next` from within the condition that sets the `forked`s' value?
E.g.

```
if (_.isPlainObject(fork.condition)) {
    if (fork.condition.value === req.form.values[fork.condition.field]) {       
        if (!_.includes(req.sessionModel.get('steps'), fork.target)) {
           next = req.baseUrl + fork.target;
           return next;
        }
    } 
}
```

Comment (easternbloc) [inline lib/base-controller.js:51]: That'll just short circuit the `_.each`

Comment (JoeChapman) [inline lib/base-controller.js:51]: I hate `_.each`

Comment (easternbloc) [inline lib/base-controller.js:51]: Git blame says no you dont

https://github.com/UKHomeOffice/hof-controllers/blame/cb47acd758962ee6ab82636399f944f044f983f8/lib/base-controller.js#L31

👯 👯 👯 👯 👯 👯
